Adds fuzz testing for demangle (#878)
google/glog: CMakeLists.txt

option (WITH_UNWIND "Enable libunwind support" ON)

cmake_dependent_option (WITH_GMOCK "Use Google Mock" ON WITH_GTEST OFF)

> set (WITH_FUZZING -fsanitize=fuzzer CACHE STRING "Fuzzing engine")

Reviewer: The option is not meant to hold compiler flags or the like. It should be a symbolic name (e.g., fuzzer). Depending on the choice the corresponding flags or link libraries should be added. Something like

```cmake
if (WITH_FUZZING STREQUAL fuzzer)
  target_compile_options (fuzz_demangle PRIVATE -fsanitize=fuzzer)
endifif (WITH_FUZZING STREQUAL fuzzer)
```

Author: Ok @sergiud 

Should be as you expect now ;-)

# Unit testing

> if (BUILD_FUZZING AND WITH_FUZZING_ENGINE)

Author: Trying something with only option `WITH_FUZZING`...

What do you expect for a `none` value ? Do not build fuzzers ? Build a static library of the fuzz target ? Have a simple driver like https://github.com/OISF/suricata/blob/master/src/tests/fuzz/onefile.c ?

> if (BUILD_FUZZING AND WITH_FUZZING_ENGINE)

Reviewer: `none` simply disables fuzzing altogether.

> if (BUILD_FUZZING AND WITH_FUZZING_ENGINE)

Author: `none` works as expected now

  add_executable (fuzz_demangle
    src/fuzz_demangle.cc
  )

>   target_link_libraries(fuzz_demangle glog ${WITH_FUZZING_ENGINE})

Author: fixing

>   target_link_libraries(fuzz_demangle glog ${WITH_FUZZING_ENGINE})

Author: The INTERFACE, PUBLIC or PRIVATE option must appear as the second argument, just after the target name.

>   target_link_libraries(fuzz_demangle glog ${WITH_FUZZING_ENGINE})

Reviewer: Yes, correct.

>   target_link_libraries(fuzz_demangle glog ${WITH_FUZZING_ENGINE})

Author: fixed

> if (BUILD_FUZZING)
>   add_executable (fuzz_demangle
>     src/fuzz_demangle.cc
>   )
>   target_link_libraries(fuzz_demangle glog $ENV{LIB_FUZZING_ENGINE})
> endif (BUILD_FUZZING)

Reviewer: I would use a CMake variable here. The caller can still use an environment variable if he chooses to do so.
```suggestion
if (BUILD_FUZZING AND WITH_FUZZING_ENGINE)
  add_executable (fuzz_demangle
    src/fuzz_demangle.cc
  )
  target_link_libraries (fuzz_demangle PRIVATE glog ${WITH_FUZZING_ENGINE})
endif (BUILD_FUZZING AND WITH_FUZZING_ENGINE)
```
```suggestion
  target_link_libraries (fuzz_demangle PRIVATE glog ${WITH_FUZZING_ENGINE})
```
`WITH_FUZZING_ENGINE` should be populated with sensible choices including one that disables fuzzing altogether (e.g., `NO` could work) which should also probably be the default.

Author: Good idea for the cmake variable `WITH_FUZZING_ENGINE`.

I am not sure I understand what you expect by `should be populated with sensible choices` (maybe I do not know cmake enough)

Reviewer: See my additional comments.

# Unit testing

if (WITH_FUZZING AND NOT WITH_FUZZING STREQUAL "none")
  add_executable (fuzz_demangle
    src/fuzz_demangle.cc
  )
  if (WITH_FUZZING STREQUAL "ossfuzz")
    set (LIB_FUZZING_ENGINE $ENV{LIB_FUZZING_ENGINE})
    target_link_libraries (fuzz_demangle PRIVATE glog ${LIB_FUZZING_ENGINE})
  elseif (WITH_FUZZING STREQUAL "libfuzzer")
    target_compile_options (fuzz_demangle PRIVATE -fsanitize=fuzzer)
    target_link_libraries (fuzz_demangle PRIVATE glog)
  else(WITH_FUZZING STREQUAL "libfuzzer")

>     message( FATAL_ERROR "Unsupported fuzzing engine." )

Reviewer: ```suggestion
    message (FATAL_ERROR "Unsupported fuzzing engine ${WITH_FUZZING}")
```

# Unit testing

if (WITH_FUZZING AND NOT WITH_FUZZING STREQUAL "none")
  add_executable (fuzz_detangle
    src/fuzz_demangle.cc
  )
  if (WITH_FUZZING STREQUAL "ossfuzz")

>     target_link_libraries (fuzz_demangle PRIVATE glog $ENV{LIB_FUZZING_ENGINE})

Reviewer: Why do we need `LIB_FUZZING_ENGINE` here? Isn't the `WITH_FUZZING` value sufficient? Either way, I would avoid using an environment variable here and instead use a CMake variable directly (drop `$ENV`). One can always initialize a CMake variable using an environment variable which is the usual pattern.

Author: oss-fuzz uses this environment variable to use whatever engine such as libfuzzer, afl++ or honggfuzz or other ones...

As you said 

> The option is not meant to hold compiler flags or the like. It should be a symbolic name (e.g., fuzzer)

So, I chose ossfuzz for this setup.
What do you think about it ?

So, I divide in 2 lines : setting the cmake variable from the environment and then using it...

Reviewer: Shouldn't we split `ossfuzz` into libfuzzer, afl++ or honggfuzz then? This would provide a much better overview of the choices we have.

Author: I do not think so because oss-fuzz does not use regular afl++

And oss-fuzz is adding more engines, you just keep the env variable `LIB_FUZZING_ENGINE`, and get these for free

Reviewer: I see. Let's then keep it this way.

  elseif (WITH_FUZZING STREQUAL "libfuzzer")

>     target_link_libraries (fuzz_demangle PRIVATE glog -fsanitize=fuzzer)

Reviewer: ```suggestion
    target_compile_options (fuzz_demangle PRIVATE -fsanitize=fuzzer)
    target_link_libraries (fuzz_demangle PRIVATE glog)
```

Author: ok

>   else

Reviewer: Incorrect syntax:
```suggestion
  else (WITH_FUZZING STREQUAL "libfuzzer")
```

Author: ok
